### PR TITLE
Add edit support

### PR DESCRIPTION
**Pull request by Sorunome**
Implements #470 

Tests for DiscordBot send and edit will follow in another PR along with some other things

### CODE REVIEW

**Comment by auscompgeek** [comment]
Looks like this needs a rebase now.

File: src/bot.ts [hunk 1]
```diff
@@ -387,15 +387,79 @@ export class DiscordBot {
...
         const chan = roomLookup.channel;
         const botUser = roomLookup.botUser;
         const embed = embedSet.messageEmbed;
+        log.verbose("============");
+        log.verbose(botUser, roomLookup);
+        log.verbose(editEventId);
+        const oldMsg = await chan.fetchMessage(editEventId);
+        if (!oldMsg) {
+            // old message not found, just sending this normally
+            embedSet.messageEmbed.description = `EDIT: ${embedSet.messageEmbed.description}`;
```
[the remark this answers is not in the document]
Comment (auscompgeek) [reply src/bot.ts]: Hm, how does this behave if we've been restarted since the old message was sent?

[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/bot.ts]: Message IDs are stored inside the DB

[the remark this answers is not in the document]
Comment (auscompgeek) [reply src/bot.ts]: I guess the only edge case then is a message that was sent whilst the bridge was down. Probably makes sense to just send the message then.

File: src/bot.ts [hunk 2]
```diff
@@ -387,15 +387,79 @@ export class DiscordBot {
...
         embedSet: IMatrixEventProcessorResult,
         opts: Discord.MessageOptions,
         roomLookup: ChannelLookupResult,
         event: IMatrixEvent,
+        editEventId: string,
     ): Promise<void> {
         const chan = roomLookup.channel;
         const botUser = roomLookup.botUser;
         const embed = embedSet.messageEmbed;
+        log.verbose("============");
```
[the remark this answers is not in the document]
Comment (Sorunome) [reply src/bot.ts]: Woops, that was an accident, leaving that in

**Comment by auscompgeek** [comment]
Hm, looks like this edits away the media message (instead of the text message) when a Discord message with a file is edited.

**Comment by Half-Shot** [comment]
Hi @Sorunome, apologies for the appalling long wait. I think you'll need to solve the conflicts now.

**Comment by ZerataX** [comment]
This works great for me and I now also don't have problems anymore with messages just getting deleted when edited on the discord side.
url preview on discord is a little annoying, but w/e.

**Comment by auscompgeek** [comment]
package.json apparently hasn't changed in this PR, so why has package-lock.json?

**Comment by Sorunome** [comment]
> package.json apparently hasn't changed in this PR, so why has package-lock.json?

because npm is stupid, lol. An `npm install` might change the package-lock.json without changing the package.json

**Comment by Sorunome** [comment]
> Edit: After a week of trying this, I actually have to stop using it because sending messages can now take hours where it usually took always <1s. that is at least using sqlite, maybe with proper postgres it's okay, but right now it is impossible for me to use it like this

That sounds 100% unrelated to this PR

**Comment by ZerataX** [comment]
> > Edit: After a week of trying this, I actually have to stop using it because sending messages can now take hours where it usually took always <1s. that is at least using sqlite, maybe with proper postgres it's okay, but right now it is impossible for me to use it like this
> 
> That sounds 100% unrelated to this PR

Yeah also not really seeing this connection but this consistenly happens for me when I check out this branch. not sure what's going on...

**Comment by Sorunome** [comment]
> > > Edit: After a week of trying this, I actually have to stop using it because sending messages can now take hours where it usually took always <1s. that is at least using sqlite, maybe with proper postgres it's okay, but right now it is impossible for me to use it like this
> > 
> > 
> > That sounds 100% unrelated to this PR
> 
> Yeah also not really seeing this connection but this consistenly happens for me when I check out this branch. And stops when I go back to develop. not sure what's going on...

If you are still on sqlite it might be time to switch to postgres. Other than that, your server resources or thelike might be overloaded. Maybe synapse is glitching out.

**Comment by auscompgeek** [comment]
I've been using this branch for a while seemingly without any performance issues.

**Comment by ZerataX** [comment]
yeah idk I probably just have some weird stuff going on with my instance anyways, since I still have problems with https://github.com/Half-Shot/matrix-appservice-discord/issues/584#issuecomment-592300286

Hmm there isn't really an easy way to reinstall and keep my bridged rooms is there? well that's offtopic anyways sorry.

**Pull request merged**